File: android-app/src/main/java/ai/aligned/ui/summarize/SummarizeViewModel.kt
package ai.aligned.ui.summarize

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import ai.aligned.data.StoryRepository
import ai.aligned.net.AlignedApi
import ai.aligned.net.dto.StoryDto
import ai.aligned.net.dto.TweetDto
import ai.aligned.net.dto.TweetInput
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * Pasted X/Twitter URL → resolve the related-tweet cluster → POST /api/summarize.
 *
 * Resolution strategy:
 *  1. Extract tweet ID from the URL via [tweetIdFromUrl].
 *  2. Look in the cached feed for a story whose tweets contain that ID. Stories
 *     on alignednews.ai are already clusters of the original tweet + its quotes
 *     and replies, so finding one story gives us the whole thread for free.
 *  3. If nothing in cache, try `/api/search` with the username from the URL.
 *  4. As a final fallback, send a single placeholder tweet so the model can
 *     still respond with something meaningful.
 */
data class SummarizeState(
    val loading: Boolean = false,
    val source: String = "",         // "Cached cluster", "Searched feed", "Direct"
    val story: StoryDto? = null,     // resolved story, if any
    val tweets: List<TweetDto> = emptyList(),
    val overview: String = "",
    val perTweet: List<String> = emptyList(),
    val error: String? = null
)

@HiltViewModel
class SummarizeViewModel @Inject constructor(
    private val api: AlignedApi,
    private val repo: StoryRepository
) : ViewModel() {

    val input: MutableState<String> = mutableStateOf("")
    private val _state = MutableStateFlow(SummarizeState())
    val state: StateFlow<SummarizeState> = _state.asStateFlow()

    fun setInput(s: String) { input.value = s }

    fun submit() {
        val url = input.value.trim()
        if (url.isBlank() || _state.value.loading) return
        val tweetId = tweetIdFromUrl(url)
        val handle = handleFromUrl(url)

        _state.value = SummarizeState(loading = true)
        viewModelScope.launch {
            runCatching {
                val resolved = resolveCluster(tweetId, handle, url)
                if (resolved.tweets.isEmpty()) {
                    error("Couldn't find any tweets for this URL. Try pasting the tweet text instead.")
                }
                val resp = api.summarize(resolved.tweets.map {
                    TweetInput(text = it.text, authorUsername = it.authorUsername)
                })
                _state.value = SummarizeState(
                    loading = false,
                    source = resolved.source,
                    story = resolved.story,
                    tweets = resolved.tweets,
                    overview = resp.overview,
                    perTweet = resp.individual
                )
            }.onFailure { e ->
                _state.update { it.copy(loading = false, error = e.message ?: "Failed to summarize") }
            }
        }
    }

    private data class Cluster(
        val source: String,
        val story: StoryDto?,
        val tweets: List<TweetDto>
    )

    private suspend fun resolveCluster(tweetId: String?, handle: String?, originalUrl: String): Cluster {
        // 1. Cached cluster
        if (tweetId != null) {
            repo.findStoryByTweetId(tweetId)?.let { story ->
                return Cluster("Found in cached feed cluster", story, story.tweets)
            }
        }
        // 2. Search by handle, then look for the tweet within those story results
        if (handle != null) {
            val searched = runCatching { api.search(handle, limit = 20) }.getOrNull()
            searched?.stories?.let { stories ->
                if (tweetId != null) {
                    val token = "/status/$tweetId"
                    for (story in stories) {
                        if (story.tweets.any { it.url.contains(token) || it.id == tweetId }) {
                            return Cluster("Resolved via /api/search", story, story.tweets)
                        }
                    }
                }
                stories.firstOrNull()?.let { story ->
                    return Cluster("Closest match by author", story, story.tweets)
                }
            }
        }
        // 3. Direct: send a single placeholder so the model can still try.
        val placeholder = TweetDto(
            id = tweetId ?: "0", text = "Tweet at $originalUrl",
            authorName = handle ?: "", authorUsername = handle ?: "",
            authorProfileImage = "", createdAt = "",
            url = originalUrl
        )
        return Cluster("Direct (tweet not in feed)", null, listOf(placeholder))
    }

    fun clear() {
        input.value = ""
        _state.value = SummarizeState()
    }
}

private val TweetIdRegex = Regex("""(?:twitter\.com|x\.com)/([^/?#]+)/status/(\d+)""", RegexOption.IGNORE_CASE)

fun tweetIdFromUrl(url: String): String? = TweetIdRegex.find(url)?.groupValues?.get(2)
fun handleFromUrl(url: String): String?  = TweetIdRegex.find(url)?.groupValues?.get(1)

File: windows-app/Aligned.App/ViewModels/SummarizeViewModel.cs
using System.Collections.ObjectModel;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using Microsoft.UI.Xaml;
using Aligned.App.Theming;
using Aligned.Core.Net;
using Aligned.Core.Net.Dto;

namespace Aligned.App.ViewModels;

/// <summary>
/// Paste an X/Twitter URL → resolve the related-tweet cluster → POST /api/summarize.
/// </summary>
public partial class SummarizeViewModel : ObservableObject
{
    private readonly AlignedApi _api = new();

    [ObservableProperty] private string _input = "";
    [ObservableProperty] private bool _isLoading;
    [ObservableProperty] private string? _errorMessage;
    [ObservableProperty] private string _source = "";
    [ObservableProperty] private string _overview = "";
    [ObservableProperty] private StoryDto? _resolvedStory;

    public ObservableCollection<string> Bullets { get; } = new();
    public ObservableCollection<TweetDto> Tweets { get; } = new();

    public string Subtitle =>
        IsLoading ? "Resolving thread…"
        : !string.IsNullOrEmpty(Overview) ? $"{Tweets.Count} tweets summarized"
        : "Paste any X / Twitter URL";

    public Visibility HintVisibility =>
        !IsLoading && string.IsNullOrEmpty(Overview) && ErrorMessage == null
            ? Visibility.Visible : Visibility.Collapsed;
    public Visibility ResultVisibility =>
        !string.IsNullOrEmpty(Overview) ? Visibility.Visible : Visibility.Collapsed;
    public Visibility ErrorVisibility =>
        !string.IsNullOrEmpty(ErrorMessage) ? Visibility.Visible : Visibility.Collapsed;
    public Visibility ClearVisibility =>
        !string.IsNullOrEmpty(Overview) || !string.IsNullOrEmpty(ErrorMessage)
            ? Visibility.Visible : Visibility.Collapsed;
    public Visibility ResolvedStoryVisibility =>
        ResolvedStory != null ? Visibility.Visible : Visibility.Collapsed;

    partial void OnIsLoadingChanged(bool value) => Notify();
    partial void OnOverviewChanged(string value) => Notify();
    partial void OnErrorMessageChanged(string? value) => Notify();
    partial void OnResolvedStoryChanged(StoryDto? value) => OnPropertyChanged(nameof(ResolvedStoryVisibility));
    partial void OnInputChanged(string value) => SubmitCommand.NotifyCanExecuteChanged();

    private void Notify()
    {
        OnPropertyChanged(nameof(Subtitle));
        OnPropertyChanged(nameof(HintVisibility));
        OnPropertyChanged(nameof(ResultVisibility));
        OnPropertyChanged(nameof(ErrorVisibility));
        OnPropertyChanged(nameof(ClearVisibility));
    }

    public bool CanSubmit => !IsLoading && !string.IsNullOrWhiteSpace(Input);

    [RelayCommand(CanExecute = nameof(CanSubmit))]
    public async Task SubmitAsync()
    {
        if (!CanSubmit) return;
        var url = Input.Trim();
        var (handle, id) = TweetLink.Parse(url);

        IsLoading = true;
        ErrorMessage = null;
        Overview = "";
        Bullets.Clear();
        Tweets.Clear();
        ResolvedStory = null;

        try
        {
            var cluster = await ResolveClusterAsync(id, handle, url);
            if (cluster.Tweets.Count == 0)
                throw new InvalidOperationException("Couldn't find any tweets for this URL.");

            ResolvedStory = cluster.Story;
            Source = cluster.Source;
            foreach (var t in cluster.Tweets) Tweets.Add(t);

            var inputs = cluster.Tweets
                .Select(t => new TweetInput(t.Text, t.AuthorUsername))
                .ToList();
            var resp = await _api.Summarize(inputs);

            Overview = resp?.Overview ?? "";
            foreach (var b in resp?.Individual ?? new()) Bullets.Add(b);
        }
        catch (Exception ex) { ErrorMessage = ex.Message; }
        finally { IsLoading = false; }
    }

    [RelayCommand]
    public void Clear()
    {
        Input = "";
        Overview = "";
        ErrorMessage = null;
        Source = "";
        Bullets.Clear();
        Tweets.Clear();
        ResolvedStory = null;
        Notify();
    }

    private async Task<Cluster> ResolveClusterAsync(string? id, string? handle, string originalUrl)
    {
        // 1. Look at the cached news payload for any tweet matching the ID.
        var news = await WarmCache.ReadStale<NewsDto>("news");
        if (id != null && news?.Stories != null)
        {
            var token = $"/status/{id}";
            foreach (var s in news.Stories)
            {
                if (s.Tweets.Any(t => t.Url.Contains(token, StringComparison.OrdinalIgnoreCase) || t.Id == id))
                    return new Cluster("Found in cached feed cluster", s, s.Tweets);
            }
        }

        // 2. Search by author handle.
        if (!string.IsNullOrEmpty(handle))
        {
            try
            {
                var searched = await _api.Search(handle, limit: 20);
                if (searched?.Stories != null)
                {
                    if (id != null)
                    {
                        var token = $"/status/{id}";
                        foreach (var s in searched.Stories)
                        {
                            if (s.Tweets.Any(t => t.Url.Contains(token, StringComparison.OrdinalIgnoreCase) || t.Id == id))
                                return new Cluster("Resolved via /api/search", s, s.Tweets);
                        }
                    }
                    var first = searched.Stories.FirstOrDefault();
                    if (first != null)
                        return new Cluster("Closest match by author", first, first.Tweets);
                }
            }
            catch { /* fall through */ }
        }

        // 3. Direct fallback: ship a single placeholder so the model can still try.
        var placeholder = new TweetDto
        {
            Id = id ?? "0",
            Text = $"Tweet at {originalUrl}",
            AuthorName = handle ?? "",
            AuthorUsername = handle ?? "",
            Url = originalUrl
        };
        return new Cluster("Direct (tweet not in feed)", null, new List<TweetDto> { placeholder });
    }

    private record Cluster(string Source, StoryDto? Story, IReadOnlyList<TweetDto> Tweets);
}
